Melodicht/tdd framework + unit testing

melodicht opened the pull request:
It is very important that we invest our time into unit testing. The unit tests will tell you if there is an error with your code without needing you to run it. Thus, it will reduce the amount of time you need to debug, giving you more time to code and refactor.

The unit tests are made possible with Jest Js, and support from Babel. To run your test, you can just run `npm run test` or `npx jest`. 

Moreover, this PR also adds a fast test-driven development (tdd) framework. Every time you save a file, the tests will automatically run. A slight inconvenience to this is that it tells you all the success. While it is not an issue now, as we have more tests, it will become quite annoying. Thus, there should be plans made to change the settings of the unit tests so that it only shows the errors. For this to work, you need to `npm run tdd` or `./tdd`.

On a side note, I also made some changes to the ESLint environments so that redundant warnings don't show up in my IDE.

EDIT: `yarn.lock` removed as `package-lock.json` already tracks dependencies.

Current goals:
- [x] Make it so that only failed tests show up in the fast tdd feedback.

melodicht committed:
Specifies more environments

To reduce redundant ESLint warnings.

melodicht committed:
Add `yarn.lock`

Yarn manages dependencies.

melodicht committed:
Update dependencies

Added further dependencies from Babel, to support Jest JS.

`gaze-run-interrupt` has also been added to support the
test-driven development (tdd) workflow.

melodicht committed:
Update dependencies

(should have done this with `package.json`, my bad)

melodicht committed:
Add support for unit testing with Jest JS

melodicht committed:
Add support for fast tdd

Run `./tdd` or `npm run tdd`, and the tests will run as you save.

aiyan commented:
Since we're using NPM, could we remove `yarn.lock` and solely track dependencies in `package-lock.json`?

melodicht committed:
Remove `yarn.lock`

melodicht committed:
Changes fast tdd so that only errors are shown

Tests that succeed would only clutter the screen.

If one would like to see them, one can run `npm run test`.

melodicht commented:
> Since we're using NPM, could we remove `yarn.lock` and solely track dependencies in `package-lock.json`?

Done!

melodicht committed:
Update dependencies

Adding TSDoc and TypeDoc

melodicht committed:
Add TypeDoc website

- To view documentation on web.

- ESLint made to ignore js files in `/doc`

melodicht committed:
Remove gaze-run-interrupt

melodicht committed:
Add `clearMocks` as default to jest config

To clear all mocks before beginning individual unit tests,
reducing tests leak.

melodicht committed:
Test client `handleCommands` and `connect`

- Add helper functions, likely to be stored in new document in the future
